### ISSUE TITLE
Payload size limits

### ISSUE THREAD

**Issue by crenz** [eclipse/hono#97]
Currently, there seems to be a payload size limit of 2 KB in the rest adaptor, and a payload size limit slightly under 8 KB in the MQTT rest adaptor – submitting a payload larger than those limits fails.

I suggest to communicate these size limits in the documentation and make them configurable.

Also, there's a larger issue behind it: Some devices might have larger payloads, like e.g. a JPEG snapshot submitted by a security camera. Users could choose to have this data inside event/telemetry messages, or to open up a “side channel” (e.g. a document store separate from Hono that the device pushes to directly), and maybe reference this separate channel from within the payload. Is this a scenario that Hono should aim to provide more explicit support for?

**Comment by sophokles73** [comment]
> I suggest to communicate these size limits in the documentation and make them configurable.

I agree, we should make this configurable.

> Some devices might have larger payloads, like e.g. a JPEG snapshot submitted by a security camera. Users could choose to have this data inside event/telemetry messages, or to open up a “side channel” (e.g. a document store separate from Hono that the device pushes to directly), and maybe reference this separate channel from within the payload. Is this a scenario that Hono should aim to provide more explicit support for?

I do not think that devices should upload large payloads to Hono's telemetry API. The API is intended for *event* size messages, i.e. several bytes up to maybe a few kilobytes. If anything larger than that needs to be published, I agree with you that the device should upload the data to some other storage system and only provide a reference to this location as part of the telemetry data. However, I am not sure how *more explicit support* for this use case should/could look like? Any suggestions?

**Comment by sophokles73** [comment]
@ppatierno 

I have tried to find an option in vertx MQTT server to limit the inbound message size but haven't bind able to find one. Is there a way to limit the message size? If not, I think it should be added to the MQTT server in order to prevent DOS attacks by uploading very large messages. WDYT?

**Comment by ppatierno** [comment]
@sophokles73 it's on my todo list digging into this because I don't know why there is this limit for now. My code has no such limit ... possibly the underlying Netty code ? or something related to the AMQP downstream path ? But in this case why a different limit for the REST and MQTT that use the same AMQP downstream path to the router ?

**Comment by crenz** [comment]
@sophokles73 

> However, I am not sure how more explicit support for this use case should/could look like? Any suggestions?

It could be a separate protocol to establish a link to a BLOB store. Possible variants:
1) BLOB attached to events: Specific properties attached to event messages could notify receivers that this particular event has additional data available under ID “y” in BLOB store “y”.
2) Registration of BLOB stores: Clients could notify the devices (in an authenticated fashion) that they are a data store capable of storing large amounts data and would like to receive data directly (via a protocol that would need to be defined).

By the way, in the concrete case I was trying out, I was posting the data as events, not as telemetry. I think in the current implementation, it does not make a difference in the REST adaptor, though.

**Comment by ppatierno** [comment]
The max 8 KB size for the MQTT adapter seems to be related to this issue opened here : 

https://github.com/vert-x3/vertx-mqtt-server/issues/10

I'll dig into it for making it something customizable,

**Comment by ppatierno** [comment]
Personally I like the idea to use the "event" path for example as a notification channel from the device to a receiver application to give information about a "large payload" (i.e. file) uploaded elsewhere.
I'm in doubt that it should be a client capability to notify devices that it's data store capable, because the data store could not belong to the client itself but being more "neutral". It's something we should thing about that because as @crenz reported it's a real scenario (i.e. snapshots sent by a camera)

**Comment by sophokles73** [comment]
I can imagine the protocol adapter handling the *offloading* of the large payload to a second tier storage (e.g. S3) and then forwarding a *reference-only* message which includes the pointer to the payload only ...

**Comment by ppatierno** [comment]
@sophokles73 you are saying that the large payload should be carried by the MQTT message and then having the adapter doing the "dirty" job ? Mhhh ... I'm not for using the MQTT protocol (but even other protocols like AMQP or even HTTP with chuncked payloads) in this way. 

Eclipse Hono shouldn't be in charge to upload file but it could provides help for that "pattern" specifying path to files in the API. The device should be in charge to send a file to a storage (i.e. S3) and the sending a simple "file notification" message to Hono (on the "event" path).

Just my 2 cents ...

**Comment by sophokles73** [comment]
@ppatierno 
I wouldn't implement this in Hono itself as a standard feature. I am only offering an opportunity for developers of *custom* protocol adapters to implement this so that we can keep payload size limitations throughout the rest of Hono. Deciding on whether using MQTT or HTTP for uploading large payloads is a good choice is then up to the implementor of the protocol adapter and the device. Note, though, that in many cases the way that the device connects and sends its data can not be (easily) changed so you need to find a way to deal with it ...

**Comment by sophokles73** [comment]
I have made the max payload size configurable in REST adapter. The default is still 2kb but it can be configured by means of the `HONO_HTTP_MAX_PAYLOAD_SIZE` environment variable.
Hono server does not support limiting the message size because I haven't yet found out how to limit message size in vertx-proton (@ppatierno maybe you can provide a pointer?).
Dispatch Router is configured to a max message size of 100kb by default.

**Comment by sophokles73** [comment]
@ppatierno 

my understanding is that we can limit the payload size in MQTT adapter once we have upgraded to vertx-mqtt-server 3.4.0, right?

**Comment by ppatierno** [comment]
@sophokles73 yes you're right. I have updated the Vert.x MQTT server CQ (https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12233) with latest 3.4.0 source code, I hope it will be accepted soon.

**Comment by ppatierno** [comment]
I have just checked the MQTT server code and unfortunately we have to wait for the 3.4.2 version (currently only as snapshot) for having limitation for message size on MQTT. With 3.4.0 I added the possibility to specify such maximum size but, the server doesn't provide to the application (adapter in this case) a way for catching exception at codec level; Netty raises an exception when maximum size is exceeded. I have added that to the 3.4.2 version only :-(

**Comment by sophokles73** [comment]
Yes, I have noticed that. However, the payload size limit still is enforced and the connection to the client is closed. I would consider this a *partial* success ;-)

BTW I would like to close this issue because we have achieved the overall goal of being able to limit the size of messages being uploaded to the adapters and the dispatch router. I will open a separate issue for defining max payload size for Hono server, though.
